statefun-sdk-js/src/core.ts

```diff
@@ -0,0 +1,329 @@
+}
+
+/**
+ * A representation of a single function.
+ * This can be created with the following object:
+ * {
+ *     typename: "foo.bar/baz",
+ *     fn(context, message) {
+ *         ...
+ *     },
+ *     specs: [..]
+ * }
+ */
+export interface FunctionOpts {
+    typename: string;
+    fn: (context: Context, message: Message) => void | Promise<void>;
+    specs?: ValueSpecOpts[]
+}
+
+/**
+ * an internal representation of function spec
+ */
+class ValueSpec implements ValueSpecOpts {
+    readonly #name: string;
+    readonly #type: Type<any>;
+    readonly #expireAfterCall: number;
+    readonly #expireAfterWrite: number;
+
+    constructor(name: string, type: Type<any>, expireAfterCall?: number, expireAfterWrite?: number) {
+        this.#name = name;
+        this.#type = type;
+        this.#expireAfterCall = expireAfterCall || -1;
+        this.#expireAfterWrite = expireAfterWrite || -1;
+    }
+
+    /**
+     * Creates a ValueSpec.
+     *
+     * @param {string} name the unique state name to use. Must be lowercase a-z or _.
+     * @param {Type} type the statefun type to associated with this state.
+     * @param {int} expireAfterCall the time-to-live (milliseconds) of this value after a call
+     * @param {int} expireAfterWrite the time-to-live (milliseconds) of this value after a write
+     * @returns {ValueSpec}
+     */
+    static fromOpts({name, type, expireAfterCall, expireAfterWrite}: ValueSpecOpts) {
+        if (isEmptyOrNull(name)) {
+            throw new Error("missing name");
+        }
+        if (!/^[_a-z]+$/.test(name)) {
+            throw new Error(`a name can only contain lower or upper case letters`);
+        }
+        if (type === undefined || type === null) {
+            throw new Error("missing type");
+        }
+        if (!Number.isInteger(expireAfterCall || -1)) {
+            throw new Error("expireAfterCall is not an integer");
+        }
+        if (!Number.isInteger(expireAfterWrite || -1)) {
+            throw new Error("expireAfterWrite is not an integer");
+        }
+        return new ValueSpec(name, type, expireAfterCall, expireAfterWrite);
+    }
+
+    /**
+     *
+     * @returns {string} the name of the this spec
+     */
+    get name() {
+        return this.#name;
+    }
+
+    /**
+     * @returns {Type} this StateFun type.
+     */
+    get type() {
+        return this.#type;
+    }
+
+    get expireAfterWrite() {
+        return this.#expireAfterWrite;
+    }
+
+    get expireAfterCall() {
+        return this.#expireAfterCall;
+    }
+}
+
+
+/**
+ * An internal representation of a function spec.
+ * A function specification has a typename, a list of zero or more declared states, and an instance of a function to invoke.
+ */
+class FunctionSpec implements FunctionOpts {
+    readonly #typename;
+    readonly #fn;
+    readonly #valueSpecs;
+
+    constructor(typename: string, fn: (context: Context, message: Message) => void | Promise<void>, specs: ValueSpec[]) {
```

Comment:
Maybe it would make sense to extract a type definition for this function to be used here and in `FunctionOpts`:

```
type ChooseAName = (context: Context, message: Message) => void | Promise<void>;
```